### PR TITLE
077 autoquant gpt fast

### PR DESCRIPTION
**Pull request by HDCharles**
Summary:
    
autoquant wasn't working for llama benchmarks for a few reasons, the main
ones being that we were doing autoquant logging on prefill not decode_one_token which is an issue since the two 
have different shapes. We also weren't torch.compiling prefill which obviated the whole point of
autoquant benchmarking torch.compiled prefill shapes.

To fix this, new functionality was needed for autoquant, we needed an
option to not automatically end logging upon a single instance of
model.forward. The flag `manual` now controls whether you
manually have to call model.finalize_autoquant() after logging is done, or
whether it happens automatically after a model forward run.

a few small other fixes were also made:
1) updated where generate.py resets cuda memory so as to not confound
   with torch.compilation memory usage
2) README updated with new numbers
3) better autoquant docstring
4) reordered benchmarks so they match whats in the README
5) cleaned up a few autoquant print multi shape print bugs

Test Plan: sh benchmarks.sh

python test_integration.py -k "test_autoquant_manual"

Reviewers:

### CODE REVIEW

File: torchao/quantization/autoquant.py [hunk 1]
```diff
@@ -479,32 +491,57 @@ def autoquant_prehook(module, args, kwargs):
...
+        # the autoquant_prehook intercepts the forward call, performs logging then
+        # does autoquantization. if model is a torch.compile wrapper, it then
+        # does the tracing/compile since the prehook is naturally followed by the normal.
+        # model run.
+        handle = model.register_forward_pre_hook(autoquant_prehook, with_kwargs=True)
+
+    # provide a method to bypass the prehook/torch.compilation
+    real_model.forward_log_only = real_model.forward
+        
+    # note the torch.compile wrapper (eval_frame) moves the assignment of any assigned
```
Comment (msaroufim) [inline torchao/quantization/autoquant.py:527]: one option is `model.compile()` instead of `torch.compile()` model which doesnt do any shenanigans with the module structure

File: test/integration/test_integration.py [hunk 2]
```diff
@@ -1197,6 +1197,43 @@ def test_autoquant_compile(self, device, dtype, m1, m2, k, n):
...
+        mod(example_input2)
+        mod.do_autoquant()
+        out2 = mod(example_input)
+        sqnr = SQNR(out, out2)
+        self.assertTrue(sqnr >= 30)
+
+        mod2 = torchao.autoquant(model, manual_do_autoquant=True)
+        mod2(example_input)
+        mod2(example_input2)
+        mod2.do_autoquant()
```
Comment (msaroufim) [inline test/integration/test_integration.py:1231]: the api names could use some work I'd expect `auto_quant` to quantize but it seems to be closer to `prepare_for_auto_quan`t and at least from this test also requires a warmup

Then there's `manual_do_autoquant` but you also have to to call `do_autoquant()`

**Review by msaroufim — CHANGES_REQUESTED** [review]

File: torchao/quantization/autoquant.py (hunk 1 above)
Comment (HDCharles) [reply torchao/quantization/autoquant.py:527]: we still want to support torch.compile though which is the more common API

File: test/integration/test_integration.py (hunk 2 above)
Comment (HDCharles) [reply test/integration/test_integration.py:1231]: its the same design as torch.compile, which doesn't actually do any compilation, it sets the model up in a way that when it gets run, it will be torch.compiled. This is the same.

If you have multiple inputs, you need a way to prevent that automatic autoquantization from happening after a model run, is there a better name that makes more sense?

Comment (msaroufim) [reply test/integration/test_integration.py:1231]: I guess yeah I'd expect `manual_do_autoquant` to be False for most users, if this is just meant to be a test utility that's a different matter

In case this is necessary then some better names could be

`mod2 = torchao.autoquant(model, quantize_on_inference=False)`

And then `mod2.quantize()`

Comment (msaroufim) [reply test/integration/test_integration.py:1231]: Steps are
1. Prepare a model for autoquantization
2. Defer quantization until all the relevant input shapes are seen
3. Once that list of shapes is seen then perform the autoquantization

File: torchao/quantization/autoquant.py [hunk 3]
```diff
@@ -479,32 +491,57 @@ def autoquant_prehook(module, args, kwargs):
...
+            return args, kwargs
+
+        # the autoquant_prehook intercepts the forward call, performs logging then
+        # does autoquantization. if model is a torch.compile wrapper, it then
+        # does the tracing/compile since the prehook is naturally followed by the normal.
+        # model run.
+        handle = model.register_forward_pre_hook(autoquant_prehook, with_kwargs=True)
+
+    # provide a method to bypass the prehook/torch.compilation
+    real_model.forward_log_only = real_model.forward
```
Comment (msaroufim) [inline torchao/quantization/autoquant.py:525]: nit for later: how to call a hf model that calls generate or decode instead of forward or call

Comment (HDCharles) [reply torchao/quantization/autoquant.py:525]: this is actually another reason for needing the manual mode, you don't have to do anything different, call huggingface as normal and then the logging will occur (the logging is on the subclass not the model hook)

then call do_autoquant or whatever at the end.

File: torchao/quantization/autoquant.py [hunk 4]
```diff
@@ -433,11 +433,31 @@ def change_autoquantizable_to_quantized(model, **kwargs):
...
+        the AutoQuantizableLinearWeight are logged
+    3) for each AutoQuantizableLinearWeight, benchmarks are run for each member of the qtensor_class_list and
+        the fastest option is picked, resulting in a highly performant model
+
+    This autoquant function performs step 1. Steps 2 and 3 can be completed by simply running the model.  
+    If `example_input` is provided, this function also runs the model. This autoquant api can handle models which have already
+    had torch.compile applied to them, in which case, once the model is run and quantized, the torch.compile process normally
+    proceeds as well.
+
+    To log multiple inputs, the user can either call model.forward_log_only rather than forward, to avoid automatic autoquantization
```
[the remark this answers is not in the document]
Comment (HDCharles) [reply torchao/quantization/autoquant.py]: updated the docstring and name to 'manual'

File: torchao/quantization/autoquant.py [hunk 5]
```diff
@@ -433,20 +433,44 @@ def change_autoquantizable_to_quantized(model, **kwargs):
...
     Args:
         model (torch.nn.Module): The model to be autoquantized.
         example_input (Any, optional): An example input for the model. If provided, the function performs a forward pass
-                                       on this input. Defaults to None.
+                                       on this input (which fully autoquantizes the model unless manual=True). Defaults to None.
         qtensor_class_list (list, optional): A list of tensor classes to be used for quantization. Defaults to DEFAULT_CLASS_LIST.
         filter_fn (callable, optional): A filter function to apply to the model parameters. Defaults to None.
         mode (list, optional): A list containing mode settings for quantization. The first element is the mode type (e.g., "interpolate"),
                                and the second element is the mode value (e.g., 0.85). Defaults to ["interpolate", .85].
+        manual (bool, optional): Whether to stop shape calibration and do autoquant after a single run (default, False) or to wait for
```
Comment (msaroufim) [inline torchao/quantization/autoquant.py:472]: nice! much clearer

File: test/integration/test_integration.py [hunk 6]
```diff
@@ -1186,17 +1186,55 @@ def test_autoquant_compile(self, device, dtype, m1, m2, k, n):
...
+        mod.do_autoquant()
+
+        out2 = mod(example_input)
+        sqnr = SQNR(out, out2)
+        self.assertTrue(sqnr >= 30)
+
+    @parameterized.expand(COMMON_DEVICE_DTYPE)
+    @unittest.skipIf(not TORCH_VERSION_AFTER_2_3, "autoquant requires 2.3+.")
+    def test_autoquant_manual(self, device, dtype):
+        if device != "cuda" and dtype != torch.bfloat16:
```
Comment (msaroufim) [inline test/integration/test_integration.py:1204]: so fp16 is skipped? how come?

File: torchao/quantization/autoquant.py [hunk 7]
```diff
@@ -433,20 +433,44 @@ def change_autoquantizable_to_quantized(model, **kwargs):
...
+    qtensor_class_list=DEFAULT_CLASS_LIST, 
+    filter_fn=None, mode=["interpolate", .85], 
+    manual=False, 
+    **aq_kwargs
+):
     """
-    Wraps the given model in an AutoQuantWrapper. If `example_input` is provided, performs a forward pass on the input.
-    Otherwise, returns the wrapped model. The AutoQuantWrapper manages cases where the model is torch-compiled by first
-    performing autoquantization on the original model and then allowing the torch.compile run/tracing to occur.
+    Autoquantization is a process which identifies the fastest way to quantize each layer of a model over some set of potential
```
Comment (msaroufim) [inline torchao/quantization/autoquant.py:445]: Could this helped quite a bit, can you also make sure it renders correctly here https://github.com/pytorch/ao/tree/main/docs/source

**Review by msaroufim — COMMENTED** [review]
Just a few more minor pieces of feedback

File: test/integration/test_integration.py [hunk 8]
```diff
@@ -1186,17 +1186,55 @@ def test_autoquant_compile(self, device, dtype, m1, m2, k, n):
...
+            self.skipTest(f"autoquant currently does not support {device}")
+        if device != "cuda" or not torch.cuda.is_available():
```
Comment (msaroufim) [inline test/integration/test_integration.py:1206]: can combine the 2 skips?

File: torchao/quantization/autoquant.py (hunk 5 above)
Comment (msaroufim) [reply torchao/quantization/autoquant.py:472]: How do you feel instead about calling this `defer_shape_calibration`?

Comment (HDCharles) [reply torchao/quantization/autoquant.py:472]: its not defering shape calibration, its elongating it

if anything it'd have to be defer_shape_calibration_end

Comment (msaroufim) [reply torchao/quantization/autoquant.py:472]: that is a mouthful lol, meh let's keep manual for now

File: torchao/quantization/autoquant.py [hunk 9]
```diff
@@ -456,17 +480,13 @@ def autoquant(model, example_input=None, qtensor_class_list=DEFAULT_CLASS_LIST,
     Example usage:
         torchao.autoquant(torch.compile(model))
         model(*example_input)
+
+        # multiple input shapes
+        torch.autoquant(model, manual=True)
+        model(*example_input1)
+        model(*example_input2)
+        model.do_autoquant()
```
Comment (jerryzh168) [inline torchao/quantization/autoquant.py:488]: why is there a autoquant and do_autoquant? seems confusing

File: torchao/quantization/autoquant.py [hunk 10]
```diff
@@ -456,17 +480,13 @@ def autoquant(model, example_input=None, qtensor_class_list=DEFAULT_CLASS_LIST,
     Example usage:
         torchao.autoquant(torch.compile(model))
         model(*example_input)
+
+        # multiple input shapes
+        torch.autoquant(model, manual=True)
```
Comment (jerryzh168) [inline torchao/quantization/autoquant.py:485]: nit: `torch` --> `torchao`

**Review by jerryzh168 — APPROVED** [review]
looks good overall, just the `do_autoquant` API feels a bit weird, I think we can make it more intuitive

File: torchao/_models/llama/generate.py [hunk 11]
```diff
@@ -206,20 +205,26 @@ def main(
...
-                2,
-                interactive=False
+                max_new_tokens,
+                interactive=False,
+                temperature=temperature,
+                top_k=top_k,
             )
+
+            # do autoquantization
+            model.do_autoquant()
```
Comment (jerryzh168) [inline torchao/_models/llama/generate.py:220]: can you also comment on why does this have to autoquant in this way

**Review by msaroufim — APPROVED** [review]

File: test/integration/test_integration.py (hunk 6 above)
Comment (HDCharles) [reply test/integration/test_integration.py]: removed this, i copied something from other tests at some point

File: torchao/_models/llama/generate.py (hunk 11 above)
Comment (HDCharles) [reply torchao/_models/llama/generate.py:220]: because we optimize for the shapes autoquant sees during shape calibration so we have to run the full generate loop, which means we need a way to manually end shape calibration and initialize benchmarking/quantization

Comment (HDCharles) [reply torchao/_models/llama/generate.py:220]: we need to do shape calibration with the actual shapes used for generate, so we setup autoquant, set it to wait until we manually end shape calibration, run the generate set to log the correct shapes, then do_autoquant to actually do the benchmarks with the shapes that we've logged

**Comment by HDCharles** [comment]
> looks good overall, just the `do_autoquant` API feels a bit weird, I think we can make it more intuitive

what terms would you use?
the difficulty is we want the api to stand on its own, but also work in a manual mode.

so

torchao.autoquant(model)

seems fine

torchao.autoquant(model, manual=True)

seems ok, maybe manual could be different? it doesn't seem super off though, the flag is 'prolonging shape calibration for multiple inputs (rather than ending after a single input and doing the benchmarks + quantization) and the user has to manually end shape calibration'

other terms could be 'manual_shape_calibration_end', 'multi_input', 'defer_finalization'....of these manual seems like the best of a bad bunch tbh

lastly there's:

model.do_autoquant()

which ends shape calibration, does benchmarking on the calibrated shapes, picks the best one and then quantizes the layers. Feels like this could be 'finalize', 'finalize_autoquant' or something along those lines. This is the step where autoquantization actually happens though so do_autoquant is literal, its just that, when manual=True, the original autoquant api is less than completely accurate, but i don't see a good way around that.
